Improvements to the showErrorDetails config

sfc-gh-lwilby opened the pull request:
## Describe your changes

This PR updates `client.showErrorDetails` to take four new settings "full", "stacktrace", "type" and "none" for more granular configuration. 

The previous true/false options are being deprecated, but are still supported in this code change. 

`full` config setting. This is equivalent to the deprecated `True` setting.
![Screenshot 2024-11-20 at 5 56 29 PM](https://github.com/user-attachments/assets/3044ff1f-2f6c-407f-875d-e715403ce9ad)

`stacktrace` config setting. This is equivalent to the deprecated `False` setting.
![Screenshot 2024-11-20 at 5 58 20 PM](https://github.com/user-attachments/assets/ada25aba-86ca-44f6-a6ac-3c8f23762a9c)

`type` config setting.
![Screenshot 2024-11-20 at 5 59 16 PM](https://github.com/user-attachments/assets/8bd29838-20dc-4154-a7b2-d271510a28ce)

`none` config setting.
<img width="924" alt="Screenshot 2024-12-04 at 6 43 53 PM" src="https://github.com/user-attachments/assets/38b7b0b2-7216-489c-ae30-aa2d6abcdcef">

## GitHub Issue Link (if applicable)

## Testing Plan

~~Explanation of why no additional tests are needed~~
- Unit Tests (JS and/or Python)
Updated existing tests and added more cases to cover the new config settings.
- E2E Tests
- Any manual testing needed?
Tested manually with global config file, command line, setting in code and environment variables. 

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-lwilby committed:
adding more granular options for error message config

sfc-gh-lwilby committed:
handle false correctly

sfc-gh-lwilby committed:
update config tests

sfc-gh-lwilby committed:
update exception tests

sfc-gh-lwilby committed:
update error_util tests

sfc-gh-lwilby committed:
update test

sfc-gh-lwilby committed:
add test for react element

sfc-gh-lwilby committed:
also test doesn't include colon

sfc-gh-lwilby committed:
remove unused import

raethlein commented on lib/streamlit/errors.py:
```diff
@@ -77,8 +77,11 @@ class MarkdownFormattedException(Error):
     """Catchall exception type for uncaught exceptions that occur during script execution."""
 
-    def __init__(self, exc):
+    def __init__(self, exc, show_message, show_type, show_trace):
```
nit: type hints are missing for the arguments

raethlein commented on frontend/lib/src/components/elements/ExceptionElement/ExceptionElement.tsx:
```diff
@@ -68,14 +68,18 @@ function ExceptionMessage({
+      if (type.length !== 0) {
+        markdown += `: `
+      }
+      markdown += `${message}`
```
super-nit: I would try to slightly rewrite this to get rid of the double-nesting and see the result string more clearly:
```typescript
let markdown = message ?? ''
if (type.length !== 0) {
  markdown = `**${type}**: ${markdown}`
}
```

but perhaps there is some catch and this doesn't work 🙂

raethlein commented on frontend/lib/src/components/elements/ExceptionElement/ExceptionElement.tsx:
```diff
@@ -68,14 +68,18 @@ function ExceptionMessage({
     <>
       <StyledMessageType>{type}</StyledMessageType>
-      {isNonEmptyString(message) ? `: ${message}` : null}
+      {type.length !== 0 && `: `}
```
nit: I think you can use normal quotes here since you don't have template literals here

raethlein commented on lib/streamlit/error_util.py:
```diff
@@ -96,13 +96,26 @@ def handle_uncaught_app_exception(ex: BaseException) -> None:
+
+    show_type = show_error_details in [FULL, STACKTRACE, TYPE, True, TRUE, FALSE, False]
+    show_trace = show_error_details in [FULL, STACKTRACE, True, TRUE, FALSE, False]
+    show_message = show_error_details in [FULL, True, TRUE]
```
nit: it looks like this could be slightly simplified to:

```python
show_message = show_error_details in [FULL, True, TRUE]
show_trace = show_message or show_error_details in [STACKTRACE, FALSE, False]
show_type = show_trace or show_error_details == TYPE
```

In my opinion this makes it a little bit easier to grasp that the different information are layered; if the message is shown, the type and trace will also always be shown; if the trace is shown, this contains that the type will be shown as well

sfc-gh-lwilby replied:
Made this change after moving the code to `marshall`.

raethlein commented on lib/streamlit/error_util.py:
```diff
@@ -96,13 +96,26 @@ def handle_uncaught_app_exception(ex: BaseException) -> None:
+    FULL = "full"
+    STACKTRACE = "stacktrace"
+    TYPE = "type"
+    # When true/false are passed in the command line they will be strings.
```
why do we have to check for `True` and `False` in [these lines](https://github.com/streamlit/streamlit/pull/9909/files#diff-a7e02f5abcde405dd67545f640b5128bd724414fbbe4f573eae30e86ea0ad1bcR110-R112) then? This comment sounds they can never be booleans

sfc-gh-lwilby replied:
When I was testing I found that if the config is set in the app, then we will have booleans and not strings.

For example, this will still result in a boolean:

```st.set_option("client.showErrorDetails", False)``` 

I can add a more in-depth comment explaining the various settings we have now, perhaps?

raethlein replied:
Thanks for the context! Yeah updating the comment with this^ info would be great for future reference

sfc-gh-lwilby replied:
I expanded this comment after moving the code into `marshall`.

raethlein commented on lib/tests/streamlit/elements/exception_test.py:
```diff
@@ -110,6 +116,30 @@ def test_uncaught_app_exception(self):
+        exception.marshall(proto, err)
+
+        for line in proto.stack_trace:
+            # assert message that could contain secret information in the stack trace
```
Is this comment correct? It sounds like we check that the secret line is in the original stack trace (but not in the proto stack trace?) I think to avoid false positives, it should also go through the original stack trace and do the positive assertion. For example, if we have a typo in the assertion string or it changes, it will pass forever (because `not in line` is always true), but would lose its purpose.

sfc-gh-lwilby replied:
Oh, thanks for catching this. I missed updating this part of the test somehow. In this test the `stack_trace` should be empty. 

I can update the original test separately as well to add the positive assertion.

sfc-gh-lwilby committed:
use double quotes and update test

sfc-gh-lwilby committed:
simplify logic

lukasmasuch commented:
An alternative approach here could be to check for the configs within the exception marshall method and apply all the changes there (all exceptions shown to the user will go through here):

https://github.com/streamlit/streamlit/blob/dbbe5fb3eaa8a6581aa598de72418ded7844576a/lib/streamlit/elements/exception.py#L76

And thereby, we can also remove the wrapping of exceptions into `UncaughtAppException` in `handle_uncaught_app_exception`. `UncaughtAppException` might be something we can fully remove from the codebase.

sfc-gh-lwilby committed:
remove usage of UncaughtAppException

sfc-gh-lwilby committed:
include False with capital f set via command line

sfc-gh-lwilby committed:
update calls to marshall

sfc-gh-lwilby committed:
fix return type of send_proto

sfc-gh-lwilby committed:
fix tests

sfc-gh-lwilby commented on lib/streamlit/runtime/app_session.py:
```diff
@@ -628,7 +628,9 @@ def _handle_scriptrunner_event_on_event_loop(
-                    msg.session_event.script_compilation_exception, exception
+                    msg.session_event.script_compilation_exception,
+                    exception,
+                    is_uncaught_app_exception=False,
```
In places calling `marshall` directly, I have added `is_uncaught_app_exception=False`. I think this should be the previous behaviour because we were only creating `UncaughtAppException` in `error_util.handle_uncaught_app_exception` so these areas calling `marshall` and enqueuing the message would not have entered that code path previously.

raethlein replied:
It might be a little bit nicer to set the default value in the args for this parameter to `False` and only update the one relevant place where it diverges from the default: 

```python
def marshall(
    exception_proto: ExceptionProto,
    exception: BaseException,
    *,
    is_uncaught_app_exception: bool = False,
)
```

sfc-gh-lwilby replied:
Yeah, I was thinking of setting it explicitly everywhere, but this probably is safer.

raethlein commented on lib/streamlit/elements/exception.py:
```diff
@@ -63,32 +63,45 @@ def exception(self, exception: BaseException) -> DeltaGenerator:
 
 
-def marshall(exception_proto: ExceptionProto, exception: BaseException) -> None:
+def send_proto(
```
nit: I guess I would name this `_send_proto` to indicate that we only use it internally and don't intend others to use it

sfc-gh-lwilby replied:
OK, I was thinking of `_` prefix as more internal to this class/file, but I'll update this.

sfc-gh-lwilby committed:
PR feedback

sfc-gh-lwilby committed:
metric trackng and test patching

sfc-gh-lwilby committed:
todo

raethlein commented on lib/streamlit/config.py:
```diff
@@ -468,8 +468,8 @@ def _logger_message_format() -> str:
         traceback show in the browser also, but they will be removed in the
         future.
```
There should be a first updated version of the doc string to reflect the new type and values (does not have to be super polished which Debbie will help with)

sfc-gh-dmatthews replied:
PR into the feature branch is here: https://github.com/streamlit/streamlit/pull/9961

raethlein commented on lib/streamlit/elements/exception.py:
```diff
@@ -56,29 +56,45 @@ def exception(self, exception: BaseException) -> DeltaGenerator:
-def marshall(exception_proto: ExceptionProto, exception: BaseException) -> None:
+# TODO(lawilby): confirm whether we want to track metrics here with lukasmasuch.
+@gather_metrics("exception")
+def _send_proto(
```
nit: I would call this `_exception` which would be aligned with other files where we do something similar for the public-facing vs. non-public facing API (e.g. in button.py)

raethlein commented on lib/streamlit/elements/exception.py:
```diff
@@ -131,10 +147,40 @@ def marshall(exception_proto: ExceptionProto, exception: BaseException) -> None:
-        exception_proto.type = type_str.replace("<class '", "").replace("'>", "")
+        show_error_details = config.get_option("client.showErrorDetails")
+
+        # Options for show error details config.
```
nit: I would change it to "Options for 'showErrorDetails' config" to make it more obvious that this is the actual name of the config

raethlein commented on lib/streamlit/elements/exception.py:
```diff
@@ -131,10 +147,40 @@ def marshall(exception_proto: ExceptionProto, exception: BaseException) -> None:
+        show_error_details = config.get_option("client.showErrorDetails")
+
+        # Options for show error details config.
+        FULL = "full"
```
I am wondering whether we should use an `Enum` class here like we do for example for [ColumnDateKind](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/elements/lib/column_config_utils.py#L47) to make the typing nicer / more structured. What do you think?

raethlein commented on lib/tests/streamlit/elements/exception_test.py:
```diff
@@ -62,11 +62,21 @@ def test_markdown_flag(self):
         proto = ExceptionProto()
-        exception.marshall(proto, StreamlitAPIException("oh no!"))
+        exception.marshall(
+            proto, RuntimeError("oh no!"), is_uncaught_app_exception=True
```
What is the reason for setting `is_uncaught_app_exception=True` here? The test does not seem to need it, but I might miss something.

sfc-gh-lwilby replied:
Since `is_uncaught_app_exception` should not affect the results of markdown I included this. I believe `RuntimeError` could be thrown from internal code or during script execution (e.g. from a library dependency). Let me know if that doesn't sound right, should I add a comment?

raethlein replied:
Ah ok, I believe in this case I would parameterize the test and this flag to make it clear that the flag does not have an impact and that the markdown-field in the proto is independent of it. Because when looking at the test right now, it looks like the flag has a meaning and is important for the tests to pass.

raethlein commented on lib/tests/streamlit/elements/exception_test.py:
```diff
@@ -150,24 +160,94 @@ def func_with_error():
-
-        assert proto.message == _GENERIC_UNCAUGHT_EXCEPTION_TEXT
-        assert proto.type == "AttributeError"
+    def test_uncaught_app_exception_default_setting_community_cloud(self):
```
nit: I think I would leave out community cloud here and refer to `False` as the generic purpose. Perhaps that `false` is used by community cloud could be a side note in the docstring

raethlein commented on lib/tests/streamlit/elements/exception_test.py:
```diff
@@ -150,24 +160,94 @@ def func_with_error():
+            exception.marshall(proto, err, is_uncaught_app_exception=True)
+
+            assert proto.message == "module 'streamlit' has no attribute 'format'"
+            assert len(proto.stack_trace)
```
nit: do you mean `assert len(proto.stack_trace) > 0`? If yes, I thing writing it out explicitly helps to understand the intention

raethlein commented on lib/tests/streamlit/elements/exception_test.py:
```diff
@@ -150,24 +160,94 @@ def func_with_error():
+            assert len(proto.stack_trace) == 0
+            assert proto.type == ""
+
+    def test_uncaught_app_exception_type_config_setting(self):
```
Is this the same as the `False` test? It looks like it and if yes, I would create a `_test...` function which is called by both or use `parameterized` with the two values

sfc-gh-lwilby replied:
This one isn't the same as `False` but I updated these tests to be focused on what is hidden/shown and parameterized the config options that should lead to that display.

raethlein commented on lib/tests/streamlit/error_util_test.py:
```diff
@@ -17,35 +17,12 @@
-    def test_uncaught_exception_show_details(self, mock_show_exception):
-        """If client.showErrorDetails is true, uncaught app errors print
-        to the frontend."""
-        with testutil.patch_config_options({"client.showErrorDetails": True}):
```
It looks like there is an equivalent test for `False` added [here](https://github.com/streamlit/streamlit/pull/9909/files#diff-5b083edb9725793b64aa5ae9f1eac9860d360633979b147e05bb6e675300f649R168), but is `True` also still covered in a test?

sfc-gh-lwilby replied:
I parameterized the tests in `exception_test` with all options.

sfc-gh-lwilby committed:
test improvements

sfc-gh-lwilby committed:
reorder tests

sfc-gh-lwilby committed:
rename to _exception

raethlein reviewed (approved):
LGTM! The feedback with the enum etc. can be a follow-up PR

sfc-gh-dmatthews committed:
client.showErrorDetails docstring (#9961)

## Describe your changes
Edit client.showErrorDetails description

---
**Contribution License Agreement**
By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.
